File: typescript/src/3_9_OfflineSignature.ts
// オフライン（オフチェーン）上で署名を集めるコード
import { PrivateKey, utils } from "symbol-sdk"
import {
  Network,
  SymbolFacade,
  descriptors,
  models,
} from "symbol-sdk/symbol"

import dotenv from "dotenv"
import { waitTxStatus } from "./waitTxStatus"

dotenv.config()

const NODE_URL = process.env.NODE_URL!
const facade = new SymbolFacade(Network.TESTNET)
const privateKeyA = new PrivateKey(process.env.PRIVATE_KEY_A!)
const accountA = facade.createAccount(privateKeyA)
const privateKeyB = new PrivateKey(process.env.PRIVATE_KEY_B!)
const accountB = facade.createAccount(privateKeyB)

// 転送トランザクション1（accountA => accountB）
const transferDescriptor1 =
  new descriptors.TransferTransactionV1Descriptor(
    accountB.address,
    [],
    "\0Hello, accountB!",
  )

// 転送トランザクション1（accountB => accountA）
const transferDescriptor2 =
  new descriptors.TransferTransactionV1Descriptor(
    accountA.address,
    [],
    "\0Hello, accountA!",
  )

const txs = [
  {
    transaction: transferDescriptor1,
    signer: accountA.publicKey,
  },
  {
    transaction: transferDescriptor2,
    signer: accountB.publicKey,
  },
]

const innerTransactions = txs.map((tx) =>
  facade.createEmbeddedTransactionFromTypedDescriptor(
    tx.transaction,
    tx.signer,
  ),
)

const innerTransactionHash =
  SymbolFacade.hashEmbeddedTransactions(innerTransactions)

const aggregateDescriptor =
  new descriptors.AggregateCompleteTransactionV2Descriptor(
    innerTransactionHash,
    innerTransactions,
  )

const txAgg = facade.createTransactionFromTypedDescriptor(
  aggregateDescriptor,
  accountA.publicKey,
  100,
  60 * 60 * 2,
  1, // 連署者数（accountBの連署が必要）
)

const signatureAgg = accountA.signTransaction(txAgg)
const jsonPayloadAgg =
  facade.transactionFactory.static.attachSignature(
    txAgg,
    signatureAgg,
  )

console.log("署名済みペイロード生成…")

const payloadAgg = JSON.parse(jsonPayloadAgg).payload

console.log("ペイロード",payloadAgg)

// （実際はこれ以降は別のコード上で実装するものだが、便宜上同じコード上に記載）
// メール等何かの方法（オフライン）でpayloadAggを送る

// ペイロードからTxの復元
console.log("ペイロードからTxの復元実施…")
const restoredTxAgg =
  models.AggregateCompleteTransactionV2.deserialize(
    utils.hexToUint8(payloadAgg),
  )

//検証を行い、改ざんされていないことを確認する
console.log("署名の検証実施…")
const responseVerify = facade.verifyTransaction(
  restoredTxAgg,
  restoredTxAgg.signature,
)

if (!responseVerify) throw new Error("署名の検証に失敗しました。")

console.log("署名の検証に成功しました。")

//accountBの連署
const cosignB = facade.cosignTransaction(
  accountB.keyPair,
  restoredTxAgg,
)

//連署者の署名追加
console.log("オフライン署名の実施…")
restoredTxAgg.cosignatures.push(cosignB)

const jsonPayloadRestoredTxAgg = JSON.stringify({
  payload: utils.uint8ToHex(restoredTxAgg.serialize()),
})

console.log("===オフライン署名したトランザクションのアナウンス===")
console.log("アナウンス開始")
const responseRestoredTxAgg = await fetch(
  new URL("/transactions", NODE_URL),
  {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: jsonPayloadRestoredTxAgg,
  },
).then((res) => res.json())

console.log("アナウンス結果", responseRestoredTxAgg)

const hashRestoredTxAgg = facade.hashTransaction(restoredTxAgg)

await waitTxStatus(
  hashRestoredTxAgg.toString(),
  NODE_URL,
  "confirmed",
)

File: python/src/3_9_offline_signature.py
# オフライン（オフチェーン）上で署名を集めるコード
import os
import json
import requests
import asyncio
from dotenv import load_dotenv
from symbolchain.CryptoTypes import PrivateKey
from symbolchain.facade.SymbolFacade import (
  SymbolFacade,
  SymbolAccount,
  Hash256,
)
from symbolchain.sc import (
  Amount,
  Signature,
  Cosignature,
  TransferTransactionV1,
  AggregateCompleteTransactionV2,
)

from wait_tx_status import wait_tx_status
from binascii import unhexlify

async def main() -> None:
  load_dotenv()

  NODE_URL: str = os.getenv("NODE_URL") or ""
  facade: SymbolFacade = SymbolFacade("testnet")

  private_key_a: str = os.getenv("PRIVATE_KEY_A") or ""
  account_a: SymbolAccount = facade.create_account(
    PrivateKey(private_key_a)
  )
  private_key_b: str = os.getenv("PRIVATE_KEY_B") or ""
  account_b: SymbolAccount = facade.create_account(
    PrivateKey(private_key_b)
  )

  network_time = requests.get(f"{NODE_URL}/node/time").json()
  receive_timestamp: int = int(
    network_time["communicationTimestamps"]["receiveTimestamp"]
  )
  deadline_timestamp: int = receive_timestamp + (
    2 * 60 * 60 * 1000
  )  # 2時間後（ミリ秒単位）

  # 転送トランザクション1(accountA=>accountB)
  transfer_tx1: (
    TransferTransactionV1
  ) = facade.transaction_factory.create_embedded({
      "type": "transfer_transaction_v1",
      "recipient_address": account_b.address,
      "mosaics": [],
      "message": b"\0Hello, accountB",
      "signer_public_key": account_a.public_key,  # 署名者の公開鍵
    })

  # 転送トランザクション2(accountA=>accountB)
  transfer_tx2: (
    TransferTransactionV1
  ) = facade.transaction_factory.create_embedded({
      "type": "transfer_transaction_v1",
      "recipient_address": account_a.address,
      "mosaics": [],
      "message": b"\0Hello, accountA!",
      "signer_public_key": account_b.public_key,  # 署名者の公開鍵
    })

  txs = [transfer_tx1, transfer_tx2]

  inner_transaction_hash: Hash256 = (
    facade.hash_embedded_transactions(txs)
  )

  # アグリゲートボンデッドトランザクションを生成
  tx_agg: (
    AggregateCompleteTransactionV2
  ) = facade.transaction_factory.create({
      "type": "aggregate_complete_transaction_v2",
      "transactions": txs,
      "transactions_hash": inner_transaction_hash,
      "signer_public_key": account_a.public_key,
      "deadline": deadline_timestamp,
    })
  tx_agg.fee = Amount(
    100 * (tx_agg.size + 1 * 104)
  )  # 連署者の署名分のサイズ （連署者 ＊ 104）を追加

  signature_agg: Signature = account_a.sign_transaction(tx_agg)
  json_payload_agg = facade.transaction_factory.attach_signature(
    tx_agg, signature_agg
  )

  print("署名済みペイロード生成…")

  # ペイロードをJSON形式に変換
  payloadAgg = str(json.loads(json_payload_agg)["payload"])

  print("ペイロード",payloadAgg)

  # メール等何かの方法（オフライン）でpayloadAggを送る
  # ここでは、payloadAggを表示することで確認する

  # ペイロードからTxの復元
  print("ペイロードからTxの復元実施…")
  restored_tx_agg = AggregateCompleteTransactionV2.deserialize(
    unhexlify(payloadAgg)
  )

  # 検証を行い、改ざんされていないことを確認する
  print("署名の検証実施…")
  response_verify = facade.verify_transaction(
    restored_tx_agg,
    restored_tx_agg.signature,
  )

  if not response_verify:
    raise Exception("署名の検証に失敗しました。")
  
  print("署名の検証に成功しました。")  

  cosignB: Cosignature = account_b.cosign_transaction(
    restored_tx_agg
  )

  # 連署者の署名追加
  print("オフライン署名の実施…")  
  restored_tx_agg.cosignatures.append(cosignB)

  json_payload_restored_tx_agg = json.dumps({
      "payload": (restored_tx_agg.serialize()).hex(),
    })

  print("===オフライン署名したトランザクションのアナウンス===")
  print("アナウンス開始")  
  response_restored_tx_agg = requests.put(
    f"{NODE_URL}/transactions",
    headers={"Content-Type": "application/json"},
    data=json_payload_restored_tx_agg,
  ).json()

  print("アナウンス結果", response_restored_tx_agg)

  hash__restored_tx_agg: Hash256 = facade.hash_transaction(
    restored_tx_agg
  )

  await wait_tx_status(
    str(hash__restored_tx_agg), NODE_URL, "confirmed"
  )


if __name__ == "__main__":
  asyncio.run(main())
